### PR TITLE
Adds Metric Logging support for GRPC workers

### PR DESCRIPTION
**Pull request by brandonh-msft**
Host-level fix to support gRPC logging of custom metric types

### Issue describing the changes in this PR

For https://github.com/Azure/azure-functions-dotnet-worker/issues/543

This adds `RpcMetric` as a gRPC message type + support to log it to ILogger in such a way that it shows in Application Insights logs as a custom metric. This provides functionality similar to what we had in in-proc languages by simply using the `.LogMetric()` ILogger extension method.

This PR encompasses the necessary changes required to complete https://github.com/Azure/azure-functions-dotnet-worker/issues/543

### Pull request checklist

* [x] My changes **do not** require documentation changes
    * [ ] Otherwise: Documentation issue linked to PR
* [ ] My changes **should not** be added to the release notes for the next release
    * [x] Otherwise: I've added my notes to `release_notes.md`
* [x] My changes **do not** need to be backported to a previous version
    * [ ] Otherwise: Backport tracked by issue/PR #issue_or_pr
* [x] I have added all required tests (Unit tests, E2E tests)

### Additional information

After completion, https://github.com/Azure/azure-functions-dotnet-worker/pull/548 should be completed to fully implement this functionality.

### CODE REVIEW

File: src/WebJobs.Script.Grpc/azure-functions-language-worker-protobuf/src/proto/FunctionRpc.proto
Comment (brettsam) [inline src/WebJobs.Script.Grpc/azure-functions-language-worker-protobuf/src/proto/FunctionRpc.proto]: I see we use `string` for RpcLog.properties above -- but it feels like `map<string, string>` makes more sense for this.

File: src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs
Comment (brettsam) [inline src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs]: We shouldn't pull Name/Value out of the properties -- `RpcMetric` has those directly on it. `Properties` should be for anything extra that you may want to pass along (such as some of the special keys that App Insights knows about).

File: src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs
Comment (brettsam) [inline src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs]: If properties is a `map<>`, I don't think we need to deserialize here.

**Review by brettsam — COMMENTED** [review]
A couple of initial comments... 

Pinging @pragnagopa for some eyes on this one as well since she's the long-time worker/grpc expert.

**Comment by brandonh-msft** [comment]
Thanks @brettsam - good catch. Didn't know there was a map type in grpc :)

File: src/WebJobs.Script.Grpc/azure-functions-language-worker-protobuf/src/proto/FunctionRpc.proto
Comment (pragnagopa) [inline src/WebJobs.Script.Grpc/azure-functions-language-worker-protobuf/src/proto/FunctionRpc.proto]: please send a PR to https://github.com/Azure/azure-functions-language-worker-protobuf adding new message

Comment (brandonh-msft) [reply src/WebJobs.Script.Grpc/azure-functions-language-worker-protobuf/src/proto/FunctionRpc.proto]: Does this replace the changes to the `.proto` files in this and the dotnet lang worker, then? or changes go in all 3 places?

Comment (pragnagopa) [reply src/WebJobs.Script.Grpc/azure-functions-language-worker-protobuf/src/proto/FunctionRpc.proto]: OK to test the changes you have with editing proto file directly in here. Please see readme in https://github.com/Azure/azure-functions-language-worker-protobuf on how to update  proto file.
- Testing with the setup you have is OK!
- When ready to check-in - Merge changes into https://github.com/Azure/azure-functions-language-worker-protobuf repo first, pull those changes into host and send separate PR - this is to make it easier to track changes to proto file that happen in host repo
- This PR should only contain code changes and no proto changes

**Comment by liliankasem** [comment]
@brandonh-msft @brettsam @pragnagopa is this PR still being worked on? Wondering if we can merge or close since it's been open since August

**Comment by brandonh-msft** [comment]
The issue is still open and the loss of `.LogMetric` is a regression from in-proc workers so IMO this should continue to be worked on until it's merged. I've done all the heavy lifting across 3 repos to make it happen & have just been waiting on approvals. It's a Draft PR because it requires that [the work in the gRPC repo](https://github.com/Azure/azure-functions-language-worker-protobuf/pull/58) be merged beforehand and I didn't want things to happen out-of-order.

**Comment by liliankasem** [comment]
> The issue is still open and the loss of `.LogMetric` is a regression from in-proc workers so IMO this should continue to be worked on until it's merged. I've done all the heavy lifting across 3 repos to make it happen & have just been waiting on approvals. It's a Draft PR because it requires that the work in the other repos be merged beforehand and I didn't want things to happen out-of-order.

Can you link the other PRs that are open? Is that something the functions team can help with or are they not owned my this team?

**Comment by brandonh-msft** [comment]
All owned by Functions and all that has needed to happen is a single button click.

**Comment by brandonh-msft** [comment]
I'm going to mark this & the dotnet worker one as 'ready for review', though, as I've moved on from my Functions-adjacent work and don't want this blocked on me.

**Review by pragnagopa — CHANGES_REQUESTED** [review]
Please send a separate PR that just updates proto file - pulling in the latest from - https://github.com/Azure/azure-functions-language-worker-protobuf/ 

This PR should only have host changes that consume changes in proto-file

**Comment by liliankasem** [comment]
> Please send a separate PR that just updates proto file - pulling in the latest from - https://github.com/Azure/azure-functions-language-worker-protobuf/
> 
> This PR should only have host changes that consume changes in proto-file

Opened #7867 to address this, will make a commit to remove proto changes from this PR

**Comment by liliankasem** [comment]
> > Please send a separate PR that just updates proto file - pulling in the latest from - https://github.com/Azure/azure-functions-language-worker-protobuf/
> > This PR should only have host changes that consume changes in proto-file
> 
> Opened #7867 to address this, will make a commit to remove proto changes from this PR

@brandonh-msft I don't have permissions to edit your branch, can you please revert all changes to the proto file

**Review by pragnagopa — DISMISSED** [review]
LGTM! @brettsam  - please review as well.

File: src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs
Comment (brettsam) [inline src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs]: This comment doesn't apply here, right?

File: src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs
Comment (brettsam) [inline src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs]: Thinking about performance here -- we turn this into a dictionary here, then later pull out two of the values and create another dictionary.

Can we instead exclude the name and value here and just create this dictionary once? Then you can pull the metric name and values straight from PropertiesMap?

File: src/WebJobs.Script.Grpc/MessageExtensions/GrpcMessageConversionExtensions.cs
Comment (brettsam) [inline src/WebJobs.Script.Grpc/MessageExtensions/GrpcMessageConversionExtensions.cs]: Don't think you need the `@` here...

File: src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs
Comment (liliankasem) [reply src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs]: great point, fixing!

**Comment by liliankasem** [comment]
Cloning this to a new branch in the host repo so we can run E2E tests, new PR:

https://github.com/Azure/azure-functions-host/pull/7956

cc: @brettsam @pragnagopa @brandonh-msft

**Pull request closed without merging**
